disk buffer panic on shutdown

### Relevant telegraf.conf

```toml
[[outputs.http]]
  namedrop      = ["internal_write"]
  url           = "https://<redacted>/instance/containers"
  method        = "POST"
  client_id     = "${CLIENT_ID}"
  client_secret = "${CLIENT_SECRET}"
  token_url     = "https://<redacted>/token"
  timeout       = "60s"

    use_batch_format = true
    content_encoding = "gzip"
    data_format      = "json"
    json_timestamp_units = "1ms"

    json_transformation = '''
{
  "requestSource":  "telegraf",
  "createInstance": true,
  "containerInstances": [{
    "originName":    metrics[0].tags.host,
    "containerName": "APMT-" & metrics[0].tags.host,
    "containerClass": "windows_client",
    "containerGroup": "clients",
    "reportEvents": $append([], metrics[tags.event_type = "report"].{
      "eventName":       tags.event_name,
      "eventObjectName": tags.object_id,
      "eventInterval":   "BASE",
      "eventData": [{"eventUtime": timestamp, "dsValues": fields}]
    }),
    "alarmEvents": $append([], metrics[tags.event_type = "alarm"].{
      "eventName":       tags.event_name,
      "eventObjectName": tags.object_id,
      "eventSeverity":   tags.severity,
      "eventData": [{"eventUtime": timestamp, "dsValues": {
        "ShortText": tags.shortText,
        "LongText":  tags.longText,
        "TodoText":  tags.todoText
      }}]
    })
  }]
}
'''
```

### Logs from Telegraf

```text
2026-07-09T12:48:40Z I! [agent] Hang on, flushing any cached metrics before shutdown
2026-07-09T12:48:41Z E! [agent] Error writing to outputs.http: Post "https://<redacted>/instance/containers": oauth2: "invalid_client" "Invalid client or Invalid client credentials"
2026-07-09T12:48:41Z I! [agent] Stopping running outputs
panic: log closed

goroutine 1 [running]:
github.com/influxdata/telegraf/models.(*DiskBuffer).readIndex(...)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:83
github.com/influxdata/telegraf/models.(*DiskBuffer).entries(0xbd3315ea580)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:73 +0x9d
github.com/influxdata/telegraf/models.(*DiskBuffer).length(...)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:69
github.com/influxdata/telegraf/models.(*DiskBuffer).Len(0xbd3315ea580)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:65 +0x7e
github.com/influxdata/telegraf/models.(*RunningOutput).BufferLength(...)
        /go/src/github.com/influxdata/telegraf/models/running_output.go:446
github.com/influxdata/telegraf/agent.(*Agent).Once(0xbd331800028, {0x7ff7f34be068?, 0xbd331844a00?}, 0xbd330ecd008?)
        /go/src/github.com/influxdata/telegraf/agent/agent.go:1074 +0x122
main.(*Telegraf).runAgent(0xbd3314aa9a0, {0x7ff7f34be068, 0xbd331844a00}, 0xde?)
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:543 +0x18e5
main.(*Telegraf).reloadLoop(0xbd3314aa9a0)
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:208 +0x26b
main.(*Telegraf).Run(0xbd3314aa9a0)
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf_windows.go:100 +0x7a5
main.runApp.func1(0xbd331b880c0)
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/main.go:283 +0xf79
github.com/urfave/cli/v2.(*Command).Run(0xbd331b8adc0, 0xbd331b880c0, {0xbd3307ca000, 0xd, 0x10})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.27.7/command.go:276 +0x7c2
github.com/urfave/cli/v2.(*App).RunContext(0xbd331018000, {0x7ff7f34bdd90, 0x7ff7fb218c80}, {0xbd3307ca000, 0xd, 0x10})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.27.7/app.go:333 +0x5a5
github.com/urfave/cli/v2.(*App).Run(...)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.27.7/app.go:307
main.runApp({0xbd3307ca000, 0xd, 0x10}, {0x7ff7f3406540, 0xbd330782020}, {0x7ff7f344cbc0, 0xbd33068d808}, {0x7ff7f344cbe8, 0xbd331aec480}, {0x7ff7f34bdc40, ...})
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/main.go:441 +0x1571
main.main()
        /go/src/github.com/influxdata/telegraf/cmd/telegraf/main.go:455 +0xe8
```

### System info

Telegraf 1.39.1 - Windows 11 AMD64

### Docker

_No response_

### Steps to reproduce

I sadly have no idea


### Expected behavior

no panic

### Actual behavior

panics upon shutdown & buffer writing

### Additional info

_No response_

--- a/models/buffer_disk.go
+++ b/models/buffer_disk.go
@@ -33,6 +33,10 @@ type DiskBuffer struct {
 	// transaction. Metrics at those offsets should not be contained in new
 	// batches.
 	mask []int
+
+	// Cache the buffer length for informatory calls to Len() e.g. when running
+	// Telegraf with --once. See https://github.com/influxdata/telegraf/issues/19248
+	closedLength *int
 }
 
 func NewDiskBuffer(id, path string, stats BufferStats, diskSync bool) (*DiskBuffer, error) {
@@ -60,6 +64,10 @@ func NewDiskBuffer(id, path string, stats BufferStats, diskSync bool) (*DiskBuff
 }
 
 func (b *DiskBuffer) Len() int {
+	if b.closedLength != nil {
+		return *b.closedLength
+	}
+
 	b.Lock()
 	defer b.Unlock()
 	return b.length()
@@ -262,9 +270,13 @@ func (b *DiskBuffer) Stats() BufferStats {
 }
 
 func (b *DiskBuffer) Close() error {
+	// Cache the length before closing the buffer
+	b.closedLength = new(b.Len())
+
 	if err := b.file.Close(); err != nil {
 		return fmt.Errorf("closing buffer failed: %w", err)
 	}
+
 	return nil
 }
 

